x/marketmap/keeper/msg_server.go

```diff
@@ -23,23 +23,83 @@ func NewMsgServer(k *Keeper) types.MsgServer {
 
-// CreateMarkets updates the marketmap by creating markets from the given message.  All updates are made to the market
-// map and then the resulting final state is checked to verify that the end state is valid.
-func (ms msgServer) CreateMarkets(goCtx context.Context, msg *types.MsgCreateMarkets) (*types.MsgCreateMarketsResponse, error) {
-	if msg == nil {
-		return nil, fmt.Errorf("unable to process nil msg")
+// UpsertMarkets wraps both Create / Update markets into a single message. Specifically
+// if a market does not exist it will be created, otherwise it will be updated. The response
+// will be a map between ticker -> updated.
+func (ms msgServer) UpsertMarkets(goCtx context.Context, msg *types.MsgUpsertMarkets) (*types.MsgUpsertMarketsResponse, error) {
+	// perform basic msg validity checks
+	ctx, err := ms.verifyMarketAuthoritiesAndExtractContext(goCtx, msg)
+	if err != nil {
+		return nil, fmt.Errorf("unable to verify market authorities: %w", err)
 	}
 
-	ctx := sdk.UnwrapSDKContext(goCtx)
+	resp := &types.MsgUpsertMarketsResponse{
+		MarketUpdates: make(map[string]bool),
+	}
 
-	params, err := ms.k.GetParams(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("unable to get marketmap params: %w", err)
+	// iterate over all markets and either create them (if no market exists), or update them
+	for _, market := range msg.Markets {
+		// check if market exists
+		exists, err := ms.k.HasMarket(ctx, market.Ticker.String())
+		if err != nil {
+			return nil, err
+		}
+
+		var eventType string
+		// if market does not exist, create it
+		if !exists {
+			err = ms.k.CreateMarket(ctx, market)
+			if err != nil {
+				return nil, err
+			}
+
+			// run hooks
+			if err = ms.k.hooks.AfterMarketCreated(ctx, market); err != nil {
+				return nil, err
+			}
+
+			resp.MarketUpdates[market.Ticker.String()] = false
+			eventType = types.EventTypeCreateMarket
+		} else {
+			err = ms.k.UpdateMarket(ctx, market)
+			if err != nil {
+				return nil, err
+			}
+
+			// run hooks
+			if err = ms.k.hooks.AfterMarketUpdated(ctx, market); err != nil {
+				return nil, err
+			}
+
+			resp.MarketUpdates[market.Ticker.String()] = true
+			eventType = types.EventTypeUpdateMarket
+		}
+
+		event := sdk.NewEvent(
+			eventType,
+			sdk.NewAttribute(types.AttributeKeyCurrencyPair, market.Ticker.String()),
+			sdk.NewAttribute(types.AttributeKeyDecimals, strconv.FormatUint(market.Ticker.Decimals, 10)),
+			sdk.NewAttribute(types.AttributeKeyMinProviderCount, strconv.FormatUint(market.Ticker.MinProviderCount, 10)),
+			sdk.NewAttribute(types.AttributeKeyMetadata, market.Ticker.Metadata_JSON),
+		)
+		ctx.EventManager().EmitEvent(event)
 	}
 
-	found := checkMarketAuthority(msg.Authority, params)
-	if !found {
-		return nil, fmt.Errorf("request signer %s does not match module market authorities", msg.Authority)
+	// validate that the new state of the marketmap is valid
+	if err = ms.k.ValidateState(ctx, msg.Markets); err != nil {
+		return nil, err
+	}
```

Comment:
seems like this would be better to have in the beginning of the function. no point in doing reads/writes if the markets in the message were bad

Reply:
oh nvm looks like this is some sort of stateful check?

Reply:
ya, this is a verification that the current-state + applied updates is valid. Agreed tho, that we don't have a nice reversion if the ultimate validation fails (we have this when applying message servers, but it kinda sucks that we have dirty state every time we apply a message)